### PR TITLE
Add methods for setting/replacing render context

### PR DESCRIPTION
**Pull request by jonbretman**
## TL;DR
* Better support for rendering with async data using `setRenderContext` and `replaceRenderContext`
* `render()` automatically called if render data is changed
* Way of providing an initial render context with `getInitialRenderContext`
* Possible to abort a render using the `shouldRender` function

## Problem 1: Rendering with async data
You want to render your component based on some data that you will fetch asynchronously.

### The current solution
```js
baustein.register('my-thing', {

    template: myTemplate,

    init: function () {
        // we need an instance property to store the render data
        // in this case it starts in a loading state
        this._renderContext = {
            loading: true
        };

        someApi().then(function (data) {
            // now we have the data we have to store a reference to it so we can return in later
            this._renderContext = data;

            // we also have to manually call render()
            this.render();
        });
    },

    getRenderContext: function () {
        return this._renderContext;
    }

});
```

Now this works fine but requires the component author to "work out" how to implement this pattern, and different components will probably end up with slightly different implementations as there is always [more than one way to skin a cat](http://s3.amazonaws.com/rapgenius/1362734102_1ebd5ac3-49ee-4eb5-81b3-f9d5332b096b.jpg).

### Proposed solution
```js
baustein.register('my-thing', {

    template: myTemplate,

    init: function () {
        someApi().then(this.replaceRenderContext.bind(this));
    },

    getInitialRenderContext: function () {
        return {
            loading: true
        };
    },

});
```

* No need for custom instance properties
* Standard way of defining the initial render state - much like `defaultOptions` for options
* Added benefit of `render()` only being called if the data returned from the API is actually different from the current render context

===

## Problem 2: Avoiding a client side render of a component that has been server rendered
You have a component that is able to render itself (e.g. it has a `template` set) but you don't want it to try to render itself if it is already rendered on the server. You also don't really want to pass all the context data used for the server render down to the client just to unnecessarily render it again.

### The current solution
One way to do this is to add an option called `rendered` to the template to tell the component that it has already been rendered and then overriding `render` to check for it.

```jinja
<div is="my-thing" rendered="true"></div>
```

```js
baustein.register('my-thing', {

    template: myTemplate,

    render: function () {
        if (this.options.rendered) {
            this.options.rendered = false;
            return this;
        }
        return baustein.Component.prototype.render.apply(this, arguments);
    }

});
```

### Proposed solution
`render()` will now **NOT** be called if a component is created from an existing DOM element.

@djmc @iprignano @skyllo @mattvagni @csharpd

### CODE REVIEW

File: test/spec/baustein.js [hunk 1]
```diff
@@ -423,23 +438,189 @@ define(['../../dist/baustein.amd.js'], function (baustein) {
...
+                        foo: 'bar'
+                    });
+
+                    var context = c.getRenderContext();
+                    context.foo = 'baz';
+
+                    expect(c.getRenderContext()).to.eql({
+                        foo: 'bar'
+                    });
                 });
```
Comment (mattvagni) [inline test/spec/baustein.js]: This test doesn't seem needed.

**Comment by mattvagni** [comment]
:+1:

**Comment by mattvagni** [comment]
Really like this.

**Comment by iprignano** [comment]
I :heart: this PR

Could it be possible to have the `shouldRender` method implicitly returning its value without having to define it in the component? Do you think it's be better to have it explicitly declared so to have less magic happening in the initialization of the component?

**Comment by iprignano** [comment]
Awesome :+1:

**Comment by mattvagni** [comment]
:+1:

**Comment by djmc** [comment]
:+1:
